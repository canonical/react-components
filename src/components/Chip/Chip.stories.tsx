import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import Chip from "./Chip";

const meta: Meta<typeof Chip> = {
  component: Chip,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof Chip>;

export const Default: Story = {
  name: "Default",

  args: {
    value: "Default",
  },
};

export const LeadValue: Story = {
  render: () => <Chip lead="Owner" value="Bob" />,
  name: "Lead-value",
};

export const Appearance: Story = {
  render: () => <Chip lead="Type" value="Positive" appearance="positive" />,
  name: "Appearance",
};

export const Dismissible: Story = {
  render: () => (
    <Chip
      lead="Owner"
      value="Bob"
      onDismiss={() => {
        console.log("onDismiss called");
      }}
    />
  ),

  name: "Dismissible",
};
