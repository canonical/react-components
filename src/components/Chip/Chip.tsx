import React from "react";
import { highlightSubString } from "../../utils";
import type { KeyboardEvent, MouseEvent } from "react";

export type Props = {
  /**
   * The lead for the chip.
   */
  lead?: string;
  /**
   * Function for handling chip div click event.
   */
  onClick?: (
    event: MouseEvent<HTMLButtonElement> | { lead: string; value: string }
  ) => void;
  /**
   * Function for handling dismissing a chip.
   */
  onDismiss?: () => void;
  /**
   * Whether the chip is selected.
   */
  selected?: boolean;
  /**
   * A substring to emphasise if it is part of the chip's value,
   * e.g. "sit" => poSITive
   */
  subString?: string;
  /**
   * Whether to wrap the value in quotation marks.
   */
  quoteValue?: boolean;
  /**
   * The value of the chip.
   */
  value: string;
};

const Chip = ({
  lead = "",
  onClick,
  onDismiss,
  quoteValue,
  selected,
  subString = "",
  value,
}: Props): JSX.Element => {
  // When user tabs over chip and presses Enter or Space key, chip will trigger
  // click functionality
  const onKeyDown = (e: KeyboardEvent<HTMLButtonElement>) => {
    // The " " value is what is returned for the spacebar
    if (e.key === " " || e.key === "Enter") {
      if (typeof onClick === "function") {
        onClick({ lead: lead, value: value });
      }
    }
  };

  const chipValue = highlightSubString(value, subString).text;

  const chipContent = (
    <>
      {lead && <span className="p-chip__lead">{lead.toUpperCase()}</span>}
      <span
        className="p-chip__value"
        dangerouslySetInnerHTML={{
          __html: quoteValue ? `'${chipValue}'` : chipValue,
        }}
      />
    </>
  );

  if (onDismiss) {
    return (
      <span className="p-chip" aria-pressed={selected}>
        {chipContent}
        {onDismiss ? (
          <button className="p-chip__dismiss" onClick={() => onDismiss()}>
            <i className="p-icon--close">Dismiss</i>
          </button>
        ) : null}
      </span>
    );
  } else {
    return (
      <button
        className="p-chip"
        aria-pressed={selected}
        onClick={onClick}
        onKeyDown={(e) => onKeyDown(e)}
      >
        {chipContent}
      </button>
    );
  }
};

export default Chip;
