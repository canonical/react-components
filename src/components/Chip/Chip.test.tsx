import { render, screen, within } from "@testing-library/react";
import React from "react";

import Chip, { Label } from "./Chip";
import userEvent from "@testing-library/user-event";

describe("Chip ", () => {
  it("renders default chip", () => {
    render(<Chip data-testid="chip" value="positive" />);
    expect(screen.getByTestId("chip")).toMatchSnapshot();
  });

  it("renders lead-value chip", () => {
    render(<Chip data-testid="chip" lead="Owner" value="Bob" />);
    expect(screen.getByTestId("chip")).toMatchSnapshot();
  });

  it("displays the value", () => {
    render(<Chip data-testid="chip" value="positive" />);
    expect(
      within(screen.getByTestId("chip")).getByText("positive")
    ).toBeInTheDocument();
  });

  it("displays the lead-value", () => {
    render(<Chip data-testid="chip" lead="Owner" value="Bob" />);
    const chip = screen.getByTestId("chip");
    expect(within(chip).getByText("OWNER")).toBeInTheDocument(); // Colon and space added with CSS
    expect(within(chip).getByText("Bob")).toBeInTheDocument(); // Colon and space added with CSS
  });

  it("displays the dismiss action", () => {
    const onDismiss = jest.fn();
    render(
      <Chip data-testid="chip" lead="Owner" value="Bob" onDismiss={onDismiss} />
    );
    expect(
      within(screen.getByTestId("chip")).getByRole("button", {
        name: Label.Dismiss,
      })
    ).toBeInTheDocument();
  });

  it("calls onDismiss when clicked", async () => {
    const onDismiss = jest.fn();
    render(
      <Chip data-testid="chip" lead="Owner" value="Bob" onDismiss={onDismiss} />
    );
    const dismissButton = within(screen.getByTestId("chip")).getByRole(
      "button",
      {
        name: Label.Dismiss,
      }
    );
    await userEvent.click(dismissButton);
    expect(onDismiss).toHaveBeenCalledTimes(1);
  });

  it("calls onKeyDown when key pressed", async () => {
    const onClick = jest.fn();
    render(
      <Chip data-testid="chip" lead="Owner" value="Bob" onClick={onClick} />
    );
    const chip = screen.getByTestId("chip");
    await userEvent.type(chip, "{enter}");
    expect(onClick).toHaveBeenCalledWith({ lead: "Owner", value: "Bob" });
  });

  it("calls onClick when clicking on the chip", async () => {
    const onClick = jest.fn();
    render(
      <Chip data-testid="chip" lead="Owner" value="Bob" onClick={onClick} />
    );
    await userEvent.click(screen.getByTestId("chip"));
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it("renders positive chip", () => {
    render(<Chip data-testid="chip" appearance="positive" value="positive" />);
    expect(screen.getByTestId("chip")).toHaveClass("p-chip--positive");
  });

  it("renders negative chip", () => {
    render(<Chip data-testid="chip" appearance="negative" value="negative" />);
    expect(screen.getByTestId("chip")).toHaveClass("p-chip--negative");
  });

  it("renders caution chip", () => {
    render(<Chip data-testid="chip" appearance="caution" value="caution" />);
    expect(screen.getByTestId("chip")).toHaveClass("p-chip--caution");
  });

  it("renders information chip", () => {
    render(
      <Chip data-testid="chip" appearance="information" value="information" />
    );
    expect(screen.getByTestId("chip")).toHaveClass("p-chip--information");
  });

  it("renders extra props", () => {
    render(
      <Chip
        data-testid="chip"
        appearance="information"
        lead="Owner"
        value="Bob"
        disabled
      />
    );
    expect(screen.getByTestId("chip")).toBeDisabled();
  });

  it("passes additional classes", () => {
    render(
      <Chip
        data-testid="chip"
        className="extra-extra"
        lead="Owner"
        value="Bob"
      />
    );
    expect(screen.getByTestId("chip")).toHaveClass("extra-extra");
  });

  it("passes additional classes to a dismissable chip", () => {
    render(
      <Chip
        data-testid="chip"
        className="extra-extra"
        lead="Owner"
        onDismiss={jest.fn()}
        value="Bob"
      />
    );
    expect(screen.getByTestId("chip")).toHaveClass("extra-extra");
  });

  it("does not submit forms when clicking on the chip", async () => {
    const onSubmit = jest.fn();
    render(
      <form onSubmit={onSubmit}>
        <Chip data-testid="chip" lead="Owner" value="Bob" />
      </form>
    );
    await userEvent.click(screen.getByTestId("chip"));
    expect(onSubmit).not.toHaveBeenCalled();
  });

  it("does not submit forms when clicking on the dismiss button", async () => {
    const onSubmit = jest.fn();
    render(
      <form onSubmit={onSubmit}>
        <Chip
          data-testid="chip"
          lead="Owner"
          onDismiss={jest.fn()}
          value="Bob"
        />
      </form>
    );
    const dismissButton = within(screen.getByTestId("chip")).getByRole(
      "button",
      {
        name: Label.Dismiss,
      }
    );
    await userEvent.click(dismissButton);
    expect(onSubmit).not.toHaveBeenCalled();
  });
});
