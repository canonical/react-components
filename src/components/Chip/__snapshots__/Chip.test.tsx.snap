// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip  renders default chip 1`] = `
<button
  className="p-chip"
  onKeyDown={[Function]}
>
  <span
    className="p-chip__value"
    dangerouslySetInnerHTML={
      Object {
        "__html": "positive",
      }
    }
  />
</button>
`;

exports[`Chip  renders lead-value chip 1`] = `
<button
  className="p-chip"
  onKeyDown={[Function]}
>
  <span
    className="p-chip__lead"
  >
    OWNER
  </span>
  <span
    className="p-chip__value"
    dangerouslySetInnerHTML={
      Object {
        "__html": "Bob",
      }
    }
  />
</button>
`;
